### DISCUSSION TITLE
annotation to explicitly select a "long string" or "text" column type

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#4396] [Design Proposals]
I believe it would be worth introducing a `@Long` or `@LongString` or `@MaxLength` annotation or something like that to make it easy for users to force the column type to `text` or `varchar(max)` or whatever, without the use of `@Lob` or JDBC-level `CLOB` APIs. On most dialects today, `@Column(length=99999999)` probably works, but it's inelegant, and likely doesn't work very consistently. And this seems like it's probably a common-enough thing to want.

**Comment by yrodiere** [comment]
Out of curiosity, what would be the equivalent in XML mapping?

I'm asking because we would have had a use for something similar in Hibernate Search, where we are currently forced to use XML mapping (the only solution for entities added by a third-party library on startup, without user configuration). We would have been able to reference a type with a name, but we wouldn't be able to use an annotation.

Though the use case in Hibernate Search is similar, but not exactly the same: we needed a binary type rather than a text type. We gave up and just used `blob` because support for binary types of undefined length without explicit indirection (non-blob) is rather inconsistent and sometimes even deprecated (DB2's `LONGVARBINARY`). But it seems to work well in at least some databases: there's `bytea` for postgresql, and I think something similar in a few other SGBDs.

**Reply by gavinking** [reply]
> we are currently forced to use XML mapping

I don't think we really have anything equivalent. I mean, there's various ways you can solve this I think: if you don't care about database portability, you just hardcode the column type. If you *do* care, then I think a `UserType` will usually work.

But we could surely add a new attribute to the XML mapping format, right?

> Though the use case in Hibernate Search is similar, but not exactly the same: we needed a binary type rather than a text type. 

Yeah, so I think we would want this to also apply to binary strings. By saying "string" I don't mean to necessarily limit it to _character_ strings.

**Reply by yrodiere** [reply]
> I don't think we really have anything equivalent. I mean, there's various ways you can solve this I think: if you don't care about database portability, you just hardcode the column type. If you do care, then I think a UserType will usually work.

Currently, without this new feature, you can just provide the type name in the XML mapping, set some high length and hope for the best. Like you would with annotation mapping, really.

> But we could surely add a new attribute to the XML mapping format, right?

Yes, I believe that would be necessary. If `@LongString` is an ad-hoc solution to this problem, something that requires specific code in the annotation "processor", then we'll need equivalent code in the XML parser.

Alternatively, I was wondering if we could provide the feature through a JDBC type. Instead of annotating the property with `@LongString`, we would annotate it with `@JdbcType(LongStringJdbcType.class)`. And in XML mapping we could refer to `LongStringJdbcType` either by its FQCN, or by some short name assigned to that type.

But maybe that's what you had in mind from the start, `@LongString` simply being an alias for `@JdbcType(LongStringJdbcType.class)`?

**Reply by gavinking** [reply]
> Alternatively, I was wondering if we could provide the feature through a JDBC type. 

It's certainly an option. I figured the dedicated annotation would be a bit more comfortable for users, but I'm not attached to it. `@JdbcType(LongStringJdbcType.class)` isn't terrible. And yeah, they're not mutually exclusive.

Dunno, I don't have strong opinion.

**Reply by sebersole** [reply]
> we would annotate it with `@JdbcType(LongStringJdbcType.class)`

XML has no facility for specifying JdbcType nor JavaType.  So same problem really

**Reply by sebersole** [reply]
> But we could surely add a new attribute to the XML mapping format, right?

For hbm.xml (which is considered deprecated btw) yes.  For orm.xml, obviously no unless we start offering a JPA XSD extension (which is actually the plan anyway.

**Reply by gavinking** [reply]
> So same problem really

Ah, OK.

**Reply by sebersole** [reply]
> > But we could surely add a new attribute to the XML mapping format, right?
> 
> For hbm.xml (which is considered deprecated btw) yes. For orm.xml, obviously no unless we start offering a JPA XSD extension (which is actually the plan anyway.

If we do decide to add a corollary to `@LongString` (or whatever) to XML mappings, then I think we need to go back and add corollaries for all of the annotations not yet supported there - and there are a lot of them.

**Comment by gavinking** [comment]
To make any sorta progress on this one we need to compile a list of the "text" or "long varchar"  types for the main platforms. This is nontrivial, since the databases are very divergent in this area.

Tell me if these are right/wrong:

- Postgres: `text`
- MySQL: `text`???? that seems a bit arbitrary...
- SQL Server: `varchar(MAX)`
- Sybase: `text`
- DB2: `varchar(32704)` ???
- Oracle: `VARCHAR2(4000)` ???? (`long` is deprecated)
- H2: `varchar(1048576)` ??
- Spanner: `string(max)`?
- Cockroach: `string`

WDYT?

**Reply by sebersole** [reply]
Up front, not a huge fan of this.. but considering we (may) want XML compatibilty, what about a "magic number" for the column length?  I.e.

```
@Column( ..., length = Integer.MAX_VALUE )
String text;
```

The only reason I suggest this is that it would work from XML as well.  Though `Integer.MAX_VALUE` would be difficult to declare in XML.  So maybe a negative value:

```
@Column( ..., length = -1 )
String text;
```

In XML

```
<column length="-1"/>
```

**Reply by gavinking** [reply]
In XML we could use `length="max"`.

**Reply by gavinking** [reply]
> In XML we could use `length="max"`.

Oh, unless the schema makes that impossible?

**Reply by sebersole** [reply]
orm.xml says this must be an int.  hbm.xml - it depends on whether you use the XSD or DTD iirc

**Reply by sebersole** [reply]
I take that back - both the hbm.xml XSD and DTD allow for String.

orm.xml does define it as int though

**Reply by sebersole** [reply]
> * Oracle: `VARCHAR2(4000)` ???? (`long` is deprecated)

Oracle does have the capability to store character data larger than 4000, though it requires a server-side setting.  4000 is definitely safest, but just for completeness.

Otherwise, that list looks correct to me

**Reply by beikov** [reply]
We already have a way to denote such types and it is the use of the JDBC types `LONGVARBINARY` and `LONGVARCHAR` which, at least in 5.x, mapped to DDL types that:

* support `getBytes`/`setBytes` and `getString`/`setString` through JDBC respectively
* are of an "unbounded" or maximum size

We also have "named" basic types that map to these JDBC types which are `image` and `text`. These names were previously usable through `@Type(..)` and still are usable in hbm.xml. I don't know about orm.xml, but I think we will need some kind of extended schema definition anyway so we could try to incorporate the type name or JDBC type code/name there. 

IMO we should change the behavior in 6.0 to match what we did in 5.x as I also tried to discuss in the Zulip chat: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Meaning.20of.20LONGVARCHAR.2FLONGVARBINARY

By the way, I have started working on this already but wanted to get consensus first before I continue. As soon as we have that, I can finish that up and provide a PR.

Also, I think it is good to use `LONGVARBINARY` and `LONGVARCHAR` as it seems that this unboundedness/maximality was the intended meaning of the type codes. Also, we did seem to have interpreted it this way in 5.x so let's continue to do so :)

**Reply by gavinking** [reply]
Well, actually, the initial code I've written does use `Types. LONGVARCHAR` because it looked convenient. 

But, on the other hand, if this is just a max-length `varchar`, it might be even more convenient to simply register it under the key `Types.VARCHAR`. In fact, many of the dialects _already have_ an existing registration that works, e.g. `SQLServerDialect` already has:

    registerColumnType( Types.VARCHAR, "varchar(max)" );

So we may not need a special JDBC type key at all.

> IMO we should change the behavior in 6.0 to match what we did in 5.x

If I recall correctly, the stuff that was there in 5 was very inconsistent across dialects (though _perhaps_ for the major dialects it was close to what I've outlined above).

> We also have "named" basic types that map to these JDBC types which are `image` and `text`. 

And that's a model I think we all agree we're trying to move away from: it's not the model which makes sense in H6. I think there's two much better options on the table here:

1. making it a special kind of "max length", or
2. @yrodiere's suggestion to just make at a `JdbcType`.

Earlier I was _leaning_ toward option 2, which has the advantage of being quite easy to implement, but now I've noticed that on many of the major platforms, this type really is just the same as the longest type registration for `varchar` (just look at the list above). So now I'm inclined to think that the correct mental model for this isn't that it's a special type, but just the "longest" `varchar`-like type, and that mental model would argue for option 1, which was more like what I was thinking about when I initially opened the discussion.

The point is that this isn't a special `Type`. It's _at most_ a special `JdbcType`, but when I opened this issue I was thinking of it as just a special length. And I now I'm leaning back in that direction.

So I don't think this is just a matter of resurrecting the code from H5.

**Reply by gavinking** [reply]
> Up front, not a huge fan of this.. but considering we (may) want XML compatibilty, what about a "magic number" for the column length?
>
>```
>@Column( ..., length = Integer.MAX_VALUE )
>String text;
>```

@sebersole My first instinct was to resist this suggestion, but after some reflection I can't see what's wrong with it.

Now, I don't love seeing `Integer.MAX_VALUE` there, but we could easily create an alias for that somewhere, I don't know where, but perhaps something like:

```java
public class VarcharJdbcType {
    public static final MAX = Integer.MAX_VALUE;
}
```

And then if I prefer I can write:

```java
@Column(length = VarcharJdbcType.MAX)
String text;
```

That's not terrible at all. There's a lot to be said for this solution:

- It works for people who're trying to use pure JPA APIs (they can just use `Integer.MAX_VALUE` as in your original example).
- It doesn't involve any new API.
- It's absurdly simple to "implement" (we only need to review the "unbounded" mappings for `Types.VARCHAR` in the major dialects).

The only question I'm really left with is would users find it easily enough? (And not see it as some sort of a hack?) We don't want them to continue misusing `@Lob`.

**Reply by gavinking** [reply]
So let me put it this way: _even if we do_ add a new annotation (or `JdbcType`) I _still_ think we should implement something like @sebersole's suggestion. Because, well, why not? It's quite a nice behavior to have for JPA-oriented users. It's a much better behavior than whatever inconsistent-across-dialects behavior we have right now.

**Reply by beikov** [reply]
Not sure if the length approach "works nicely" for XML mappings, as one would have to spell out `Integer.MAX_VALUE` which brings up questions for readers of the mapping i.e. `Why this value?`. Apart from that, it would IMO also be great if we support generating check constraints for columns even if they use these "unbounded/max DDL types". So I would prefer not to use the length to indicate the use of this DDL type.

Apart from that, for the user model, I agree that a dedicated annotation is great.

**Reply by gavinking** [reply]
A slightly refined version of this suggestion would be something like:

```java
public class VarcharJdbcType {
    public static final int LONG = Short.MAX_VALUE; \\2^16-1
    public static final int MAX = Integer.MAX_VALUE;
}
```

So that `LONG` would get you `text` on MySQL and `MAX` would get you `longtext`.

**Reply by gavinking** [reply]
@beikov as discussed above, `length='max'` would work in `hbm.xml`, but sure, `orm.xml` would be left out in the cold and ..... I guess I don't care much? The workaround for people with `orm.xml` is: pick a really big number, say `length='50000'`. That's not terrible either.

**Reply by beikov** [reply]
Well, orm.xml is going away at some point so we have to care about orm.xml users. Also, I'm not a fan of using "magic numbers".

**Reply by beikov** [reply]
I meant hbm.xml is going away

**Reply by gavinking** [reply]
Check this:

```java
public class Length {
    public static final int DEFAULT = 255;
    public static final int LONG16 = Short.MAX_VALUE;
    public static final int LONG32 = Integer.MAX_VALUE;
}
```

Now _that_ is nice. Because on some platforms (MySQL) there is a distinction between strings whose length fits in a 16-bit integer and those which don't. And since all Java `String`s fit in a 32-bit integer, we never have to contemplate anything longer (and the same goes for most of the databases).

These aren't magic numbers. These are actual meaningful values representing real limits to the datatypes.

    @Column(length=Length.LONG16)

doesn't read like any sort of hack to me.

**Reply by gavinking** [reply]
> Also, I'm not a fan of using "magic numbers".

Sure, I always avoid (perhaps irrationally) the use of `MAX_VALUE` to mean infinity.

But that's not what this is. `Integer.MAX_VALUE` _is the literal maximum length_ of a string in Java, and the maximum length of a `longtext` on MySQL. It's not a number we're pulling out of our butts. It's the real actual physical limit.

**Reply by sebersole** [reply]
> Also, I'm not a fan of using "magic numbers".

I literally started that suggestion with "not a huge fan of this.. but considering we (may) want XML compatibilty, what about a "magic number"" lol

Literally no-one is a "fan" of magic numbers ;)

**Reply by sebersole** [reply]
> Because on some platforms (MySQL) there is a distinction between strings whose length fits in a 16-bit integer and those which don't. And since all Java `String`s fit in a 32-bit integer, we never have to contemplate anything longer (and the same goes for most of the databases).

I assume the assumption is that these define a minimum size and we can up-size as needed, then fine.  

I do still consider the XML case since @yrodiere brought it up.  `hbm.xml` is simple - its a String and we can support any "magic" value we want there, e.g.

```
<column ... length="max"/>
<column ... length="16bit"/>
<column ... length="32bit"/>
```

But that still leaves the question of `orm.xml` - whether we want to allow this kind of distinction there and, if so, how we can have users indicate this via a simple int value.  And I think we all agree that hand-coding the actual value of those constants is not user-friendly at best.

But like @gavinking said, these options are not necessarily mutually exclusive

**Reply by gavinking** [reply]
> these options are not necessarily mutually exclusive

Right, I mean `LongVarbinaryJdbcType` already exists in H6, and after cleaning up the dialect support for `LONGVARCHAR` it will also work nicely cross platforms. (In H5 some dialects didn't even have sensible mappings for `LONGVARCHAR` IIRC.)

So `@JdbcType(LongVarbinaryJdbcType.class)` is one option that annotation users have.

However it seems to me that, `@Column(length=LONG32)` is nicer in certain ways.

In XML, we can interpret `-1` or `32bit` or whatever as meaning the same thing. That's just a question for the XML binder.

**Reply by gavinking** [reply]
I forgot: `@JdbcTypeCode( Types.LONGVARCHAR )` is also a perfectly good option, and works fine with the work I've already done.

**Comment by sebersole** [comment]
@gavinking I agree that this is not "terrible":

```
@Column(length = VarcharJdbcType.MAX)
String text;
```

But I do think it somewhat defeats the purpose of the suggestion which is to allow seamless functionality with XML mappings.  As you even pointed out, hand-coding that value in XML is error prone and, well, fugly.

> as one would have to spell out `Integer.MAX_VALUE`

Yep ^^

> questions for readers of the mapping i.e. Why this value?

I mean, that's the general point wrt magic numbers.  Nothing really unique about it here

> The workaround for people with orm.xml is: pick a really big number, say length='50000'. That's not terrible either.

Which is why I reversed course earlier and went with a negative number.

> doesn't read like any sort of hack to me.

I wouldn't say hack... but how do users mapping with XML achieve this?

**Reply by gavinking** [reply]
We could also allow a negative number.

**Reply by sebersole** [reply]
Do we allow 2?  One for 16 and one for 32 bit representations?

:grin:

**Reply by gavinking** [reply]
ffff, I mean, in principle yeah but....  ugh

**Reply by gavinking** [reply]
I mean, `length='0'` and `length='-1'`. :-/

**Comment by sebersole** [comment]
> We already have a way to denote such types and it is the use of the JDBC types LONGVARBINARY and LONGVARCHAR

In many ways I like and agree with this, but 2 concerns:

1. this is viable for Hibernate-specific mappings (hbm.xml) and Hibernate-specific annotations.  however, neither JPA annotations nor orm.xml allow specifying a JDBC type code.  so how does this work in those cases?
2. in a sense it is confusing when I specify `LONGVARCHAR` e.g. and use a database that does in fact have an actual corresponding datatype (such as Oracle), but Hibernate decides to use a different datatype.  not saying wrong per-se, but "surprising" from the user perspective.

**Reply by sebersole** [reply]
> this is viable for Hibernate-specific mappings (hbm.xml) and Hibernate-specific annotations. however, neither JPA annotations nor orm.xml allow specifying a JDBC type code. so how does this work in those cases?

To play my own devil's advocate, we could look at this like "JPA does not define this capability, so we should not really care about this in strict JPA mapping".  But that is a bit of a straw-man argument considering we generally push users to prefer annotations (most of which are JPA defined) over XML...

**Comment by sebersole** [comment]
All told, I say this... they all make sense in their own right, each with pros and cons.  I think it makes sense to support multiple of them.

For annotation mappers, who are willing to use Hibernate annotations, I think `@JdbcTypeCode( Types.LONGVARCHAR )` is a great option that requires no real work on our part other than to change up some Dialects in terms of which `JdbcType` they register for that code.  I also think any of these are great:

```
@Long
String theString;

@Medium
String theString;

@Long( MAX )
String theString;

@Long( MEDIUM )
String theString;

@Long( BIT16 )
String theString;

@Long( BIT32 )
String theString;
```

They all really describe the same thing.  max/medium versus 32bit/16bit really target on different levels of user.

---- 

For `hbm.xml` mappers, I don't much care if we recognize `"max"`, `"-1"`, `"32bit"`, etc.  It's really all just different shades of lipstick for the pig.

----

For pure JPA mappers (same for annotations and `orm.xml` since their DTD defines this as `int` explicitly) - I think we have to support magic numbers or something like the suggested `Length` constants (or both).  It has to be int-based so that is pretty limiting in terms of possible solutions.  I will say however that users who are unwilling to use Hibernate annotations are probably even more unwilling to use a Hibernate constant just to avoid using a Hibernate annotation.  Granted, we could document the effect of using `Short.MAX_VALUE` and `Integer.MAX_VALUE` in the user-guide and call it a day..

**Reply by gavinking** [reply]
> I think it makes sense to support multiple of them.

Right, I reached the exact same conclusion.

**Reply by gavinking** [reply]
>  I will say however that users who are unwilling to use Hibernate annotations are probably even more unwilling to use a Hibernate constant just to avoid using a Hibernate annotation.

Right, but those people can *easily* just copy the `Length` class into their own project.

> Granted, we could document the effect of using `Short.MAX_VALUE` and` Integer.MAX_VALUE` in the user-guide and call it a day.

Sure, a big part of this is documenting stuff so that people stop abusing `@Lob`.

**Reply by sebersole** [reply]
> Right, but those people can _easily_ just copy the `Length` class into their own project.

So they don't want to use Hibernate annotations; they don't want to use Hibernate classes; but they would be willing to copy a Hibernate class into their application for a feature that only Hibernate supports?

**Reply by gavinking** [reply]
Believe me, that is *exactly* the sort of slightly-irrational thing people are often prepared to do. :-)

**Reply by sebersole** [reply]
Sadly, you are probably not too far off :rofl:

**Comment by sebersole** [comment]
> Sure, a big part of this is documenting stuff so that people stop abusing @Lob.

+10000
